Mathlib/Data/Nat/ModEq.lean

```diff
@@ -530,4 +543,129 @@ theorem odd_mod_four_iff {n : ℕ} : n % 2 = 1 ↔ n % 4 = 1 ∨ n % 4 = 3 :=
+
+variable (a s : ι → ℕ)
+
+/-- The natural number less than `(l.map s).prod` congruent to
+`a i` mod `s i` for all  `i ∈ l`. -/
+def chineseRemainderOfList : (l : List ι) → l.Pairwise (Coprime on s) →
+    { k // ∀ i ∈ l, k ≡ a i [MOD s i] }
+  | [],     _  => ⟨0, by simp⟩
+  | i :: l, co => by
+    have : Coprime (s i) (l.map s).prod := coprime_list_prod_right_iff.mpr
+      (by simp; intro j hj; exact (List.pairwise_cons.mp co).1 j hj)
+    have ih := chineseRemainderOfList l co.of_cons
+    have k := chineseRemainder this (a i) ih
+    exact ⟨k, by
+      simp [k.prop.1]; intro j hj
+      exact ((modEq_list_prod_iff' co.of_cons).mp k.prop.2 j hj).trans (ih.prop j hj)⟩
+
+@[simp] theorem chineseRemainderOfList_nil :
+    (chineseRemainderOfList a s [] List.Pairwise.nil : ℕ) = 0 := rfl
+
+theorem chineseRemainderOfList_lt_prod (l : List ι)
+    (co : l.Pairwise (Coprime on s)) (hs : ∀ i ∈ l, s i ≠ 0) :
+    chineseRemainderOfList a s l co < (l.map s).prod := by
+  cases' l with i l <;> simp
+  · simp [chineseRemainderOfList]
+    have : Coprime (s i) (l.map s).prod := coprime_list_prod_right_iff.mpr
+      (by simp; intro j hj; exact (List.pairwise_cons.mp co).1 j hj)
+    exact chineseRemainder_lt_mul this (a i)
+      (chineseRemainderOfList a s l co.of_cons)
+        (hs i (by simp)) (by simp; intro j hj; exact hs j (by simp [hj]))
+
+theorem chineseRemainderOfList_modEq_unique (l : List ι)
+    (co : l.Pairwise (Coprime on s)) {z} (hz : ∀ i ∈ l, z ≡ a i [MOD s i]) :
+    z ≡ chineseRemainderOfList a s l co [MOD (l.map s).prod] := by
+  induction' l with i l ih
+  · simp [modEq_one]
+  · simp [chineseRemainderOfList]
+    have : Coprime (s i) (l.map s).prod := coprime_list_prod_right_iff.mpr
+      (by simp; intro j hj; exact (List.pairwise_cons.mp co).1 j hj)
+    exact chineseRemainder_modEq_unique this
+      (hz i (by simp)) (ih co.of_cons (fun j hj => hz j (by simp [hj])))
+
+theorem chineseRemainderOfList_perm {l l' : List ι} (hl : l.Perm l')
+    (hs : ∀ i ∈ l, s i ≠ 0) (co : l.Pairwise (Coprime on s)) :
+    (chineseRemainderOfList a s l co : ℕ) =
+    chineseRemainderOfList a s l' (co.perm hl coprime_comm.mpr) := by
+  let z := chineseRemainderOfList a s l' (co.perm hl coprime_comm.mpr)
+  have hlp : (l.map s).prod = (l'.map s).prod := List.Perm.prod_eq (List.Perm.map s hl)
+  exact (chineseRemainderOfList_modEq_unique a s l co (z := z)
+    (fun i hi => z.prop i (hl.symm.mem_iff.mpr hi))).symm.eq_of_lt_of_lt
+      (chineseRemainderOfList_lt_prod _ _ _ _ hs)
+      (by rw [hlp]
+          exact chineseRemainderOfList_lt_prod _ _ _ _
+            (by simpa [List.Perm.mem_iff hl.symm] using hs))
+
+/-- The natural number less than `(m.map s).prod` congruent to
+`a i` mod `s i` for all  `i ∈ m`. -/
+def chineseRemainderOfMultiset {m : Multiset ι} :
+    m.Nodup → (∀ i ∈ m, s i ≠ 0) → Set.Pairwise {x | x ∈ m} (Coprime on s) →
+    { k // ∀ i ∈ m, k ≡ a i [MOD s i] } :=
```

Comment:
Again, would it be better to separate out the data that's the output from the proof it works?